validator/sawtooth_validator/journal/consensus/poet/poet_consensus/poet_block_publisher.py

```diff
@@ -0,0 +1,279 @@
+        # have a public key, we need to create signup information and create a
+        # transaction to add it to the validator registry.
+        elif PoetBlockPublisher._poet_public_key is None:
+            self._register_signup_information(block_header=block_header)
+
+        # Create a list of certificates for the wait timer.  This seems to have
+        # a little too much knowledge of the WaitTimer implementation, but
+        # there is no use getting more than
+        # WaitTimer.certificate_sample_length wait certificates.
+        certificates = \
+            utils.build_certificate_list(
+                block_header=block_header,
+                block_cache=self._block_cache,
+                poet_enclave_module=self._poet_enclave_module,
+                maximum_number=WaitTimer.certificate_sample_length)
+
+        # We need to create a wait timer for the block...this is what we
+        # will check when we are asked if it is time to publish the block
+        self._wait_timer = \
+            WaitTimer.create_wait_timer(
+                poet_enclave_module=self._poet_enclave_module,
+                validator_address=block_header.signer_pubkey,
+                certificates=list(certificates))
+
+        LOGGER.info('Created wait timer: %s', str(self._wait_timer))
+
+        return True
+
+    def check_publish_block(self, block_header):
+        """Check if a candidate block is ready to be claimed.
+
+        Args:
+            block_header (BlockHeader): The block header for the candidate
+                block that is checked for readiness for publishing.
+        Returns:
+            Boolean: True if the candidate block should be claimed. False if
+            the block is not ready to be claimed.
+        """
+
+        # Only claim readiness if the wait timer has expired
+        return self._wait_timer.has_expired(now=time.time())
+
+    def finalize_block(self, block_header):
+        """Finalize a block to be claimed. Provide any signatures and
+        data updates that need to be applied to the block before it is
+        signed and broadcast to the network.
+
+        Args:
+            block_header (BlockHeader): The block header for the candidate
+                block that needs to be finalized.
+        Returns:
+            Boolean: True if the candidate block good and should be generated.
+            False if the block should be abandoned.
+        """
+        # To compute the block hash, we are going to perform a hash using the
+        # previous block ID and the batch IDs contained in the block
+        hasher = hashlib.sha256(block_header.previous_block_id.encode())
+        for batch_id in block_header.batch_ids:
+            hasher.update(batch_id.encode())
+        block_hash = hasher.hexdigest()
+
+        # We need to create a wait certificate for the block and then serialize
+        # that into the block header consensus field.
+        wait_certificate = \
```

Comment:
currently errors in the publisher will halt the publisher thread in effect stopping the validator. Any exceptions raised via consensus interface are considered fatal to the validator. You should handle expected errors internally and translate them to the expected return code if they are non fatal.